docs/src/main/sphinx/develop/connectors.rst

```diff
@@ -297,6 +301,233 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
 error code and an actionable message, to let users know how to write a valid
 query.
 
+.. _predicate-pushdown:
+
+Predicate pushdown
+^^^^^^^^^^^^^^^^^^
+
+When executing a query with a ``WHERE`` clause, the query plan may
+contain a ``Filter`` operation.
+
+When the query plan contains a ``Filter`` operation, the Trino engine
+tries to optimize the query by pushing down the predicate constraint
+into the connector by calling the ``applyFilter`` method of the
+connector metadata service. This method receives a table handle with
+all optimizations applied thus far, and should return either
+``Optional.empty()`` or a response with a new table handle derived from
+the old one. The constraint which will eventually be applied on the
+underlying source level is accumulated over multiple calls in a
+``ConnectorTableHandle``.  Once the query actually runs,
+``ConnectorRecordSetProvider`` or ``ConnectorPageSourceProvider``
+use whatever optimizations were pushed down to ``ConnectorTableHandle``.
+
+The query optimizer may call ``applyFilter`` for a single query multiple times,
+as it searches for an optimal query plan. It's important that connectors
+return ``Optional.empty()`` from ``applyFilter`` if they can't apply the
+constraint for this invocation even if they support ``WHERE`` clause pushdown
+in general. It's also important to return ``Optional.empty()`` if the
+constraint has already been applied by some previous invocation.
+
+A constraint contains:
+
+* a ``TupleDomain`` summary of possible values (or ranges) for different
+  columns,
+* an expression for pushing down function calls,
+* a map of assignments from variables in the expression to columns,
+* (optional) a predicate which tests a map of columns and their values,
+  it cannot be held on to after the ``applyFilter`` call returns
+* (optional) a set of columns the predicate depends on, must be present
+  if predicate is present.
+
+A ``TupleDomain`` defines a mapping between columns and their domains.
+A ``Domain`` is either a list of possible values or a list of ranges.
+It also contains information about nullability.
+
+If both a predicate and a summary are available, handling the predicate
+is preferred over handling the summary, as the predicate is guaranteed
+to be more strict in filtering out values.
+However it's not possible to store a predicate in the table handle and use
+it later, as the predicate cannot be held on to after the ``applyFilter``
+call returns. This means a predicate can only be used for immediate
+filtering of entire Hive partitions/DB shards and is not actually pushed down.
```

Comment:
This guide should not refer to any specific connectors. What does `immediate filtering` mean? `applyFilter` doesn't handle any data, other services do, like the split manager and record set provider (or page source provider).

Reply:
By `immediate filtering` I meant filtering that occurs inside `applyFilter`, since we can't store the predicate in a `TableHandle`. How this would be done depends on the DB implementation - in Hive connector we query the HMS for partitions, filter them using the predicate, and store the resulting list in `HiveTableHandle`. I guess sharded DBs could use a similar approach if their API allows querying for a list of shard keys.

Reply:
This is an implementation detail that we shouldn't include in the guide. I'm not entirely sure if `applyFilter` should be doing time-consuming operations like talking to external services, since query planning should be done as quickly as possible. Instead, maybe it's better to do it in the split manager, using the table handle returned here.

It's enough to say that predicate objects cannot be reused outside of this method, we don't need to go into details about how to deal with it.